code/datums/components/hallucinations.dm

```diff
@@ -0,0 +1,537 @@
+	var/pattern2 = list(0,1,0,0, 0,0,1,0, 1,0,0,0, 0,0,0,1, 0,0,0,0)
+
+	do_mob_tick(mob, mult)
+		if(parent_mob.client && (current_color_pattern == 0 || probmult(20))) //trippy colours
+			if(src.current_color_pattern == 1)
+				parent_mob.client.animate_color(pattern2, time=40, easing=SINE_EASING)
+				src.current_color_pattern = 2
+			else
+				parent_mob.client.animate_color(pattern1, time=40, easing=SINE_EASING)
+				src.current_color_pattern = 1
+		..()
+
+	UnregisterFromParent()
+		. = ..()
+		UnregisterSignal(parent, COMSIG_LIVING_LIFE_TICK)
+		if(parent_mob?.client)
+			animate(parent_mob.client, color = null, time = 2 SECONDS, easing = SINE_EASING)
+
+
+//#########################################################
+//                    RANDOM SOUNDS
+//#########################################################
+
+/// Random sound - play a sound from a list with a prob per life tick
+/datum/component/hallucination/random_sound
+	var/list/sound_list
+	var/sound_prob = 10
+
+	Initialize(timeout=30, sound_list=null, sound_prob=10)
+		.=..()
+		if(. == COMPONENT_INCOMPATIBLE || length(sound_list) == 0)
+			return .
+		src.sound_list = sound_list
+		src.sound_prob = sound_prob
+
+
+	do_mob_tick(mob, mult)
+		if(probmult(src.sound_prob))
+			var/atom/origin = parent_mob.loc
+			var/turf/mob_turf = get_turf(parent_mob)
+			if (mob_turf)
+				origin = locate(mob_turf.x + rand(-10,10), mob_turf.y + rand(-10,10), mob_turf.z)
+			//wacky loosely typed code ahead
+			var/datum/hallucinated_sound/chosen = pick(src.sound_list)
+			if (istype(chosen)) //it's a datum
+				chosen.play(parent_mob, origin)
+			else //it's just a path directly
+				parent_mob.playsound_local(origin, chosen, 100, 1)
+		. = ..()
+
+	CheckDupeComponent(timeout, sound_list, sound_prob)
+		if(sound_list ~= src.sound_list) //this is the same hallucination, just update timeout and prob
+			if(timeout == -1)
+				src.ttl = timeout
+			else if(src.ttl != -1)
+				src.ttl = world.time + timeout SECONDS //reset timeout
+			src.sound_prob = sound_prob
+			return TRUE //no duplicate
+		else
+			return FALSE //create a new hallucination
+
+//#########################################################
+//                    RANDOM IMAGE
+//#########################################################
+
+/// Random image - hallucinate an image on a visible tile with prob per life tick
+/datum/component/hallucination/random_image
+	var/list/image_list
+	var/image_prob = 10
+	var/image_time = 20 SECONDS
+
+	Initialize(timeout=30, image_list=null, image_prob=10, image_time=20 SECONDS)
+		.=..()
+		if(. == COMPONENT_INCOMPATIBLE || length(image_list) == 0)
+			return .
+		src.image_list = image_list
+		src.image_prob = image_prob
+		src.image_time = image_time
+
+	do_mob_tick(mob, mult)
+		if(probmult(image_prob))
+			//pick a non dense turf in view
+			var/list/turf/potentials = list()
+			for(var/turf/T in view(parent_mob))
+				if(T.density == 0)
```

Comment:
```suggestion
				if(!T.density)
```
fairly sure `!density` is more common since we can treat it like a bool

Reply:
copy pasting old code :waa: